Support effective watermark thresholds in node stats API

howardhuanghua opened the pull request:
Adds to the `fs` component of the node stats API some additional values
indicating the disk watermarks that are currently in effect.

Relates #106676

DaveCTurner commented on server/src/main/java/org/elasticsearch/node/NodeConstruction.java:
```diff
@@ -969,7 +969,7 @@ record PluginServiceInstances(
             settings,
             threadPool,
-            new MonitorService(settings, nodeEnvironment, threadPool),
+            new MonitorService(settings, clusterService.getClusterSettings(), nodeEnvironment, threadPool),
```
This uses the `DiskThresholdSettings` according to the local node, but that's not necessarily correct, it's the settings on the elected master node which matter. They'll be the same if the settings are set via `PUT _cluster/settings`, but different if set in `elasticsearch.yml`.

DaveCTurner replied:
If the settings are not there in `_cluster/settings` then they will come from the `elasticsearch.yml` _on the elected master_. Here you're looking at the settings from the `elasticsearch.yml` _on the local node_, and these things may be different.

howardhuanghua replied:
Hi @DaveCTurner , sorry I might misunderstanding the key point.
>then they will come from the elasticsearch.yml on the elected master.

It should not come from elected master's yml? It should come from the yml of the local node itself?
>Here you're looking at the settings from the elasticsearch.yml on the local node

The above `DiskThresholdSettings` is construncted based on `settings` and `clusterSettings` in above FsService.java L47:
`final FsProbe probe = new FsProbe(nodeEnvironment, new DiskThresholdSettings(settings, clusterSettings));`
So it could evaluate static configure and dynamic cluster settings.

I have done a real cluster integration test:
1. Setup a two nodes cluster, both have data and master role.
2. No dynamic cluster settings, and set one node's watermark as `70%, 75%, 80%`, another node keep default. Get node stats API result, they are different:
```
 "nodes" : {
    "oKcA8bgMSCeUK1vb0cOyfA" : {
      "fs" : {
        "timestamp" : 1712733138988,
        "total" : {
          "total_in_bytes" : 21003583488,
          "free_in_bytes" : 18922135552,
          "available_in_bytes" : 18261114880
        },
        "data" : [
          {
            "path" : "/data1/elasticsearch-8.13.0/data",
            "mount" : "/data1 (/dev/vdb)",
            "type" : "ext4",
            "total_in_bytes" : 21003583488,
            "free_in_bytes" : 18922135552,
            "available_in_bytes" : 18261114880,
            "effective_low_watermark" : "70%",
            "effective_high_watermark" : "75%",
            "effective_flood_stage" : "80%"
          }
        ],
...
   "data" : [
          {
            "path" : "/data1/elasticsearch-8.13.0/data",
            "mount" : "/data1 (/dev/vdb)",
            "type" : "ext4",
            "total_in_bytes" : 21003583488,
            "free_in_bytes" : 18706587648,
            "available_in_bytes" : 18045566976,
            "effective_low_watermark" : "85%",
            "effective_high_watermark" : "90%",
            "effective_flood_stage" : "95%"
          }

```
3. Update cluster settings:
```
PUT localhost:9200/_cluster/settings
{
  "persistent": {
    "cluster.routing.allocation.disk.watermark.low": "50%",
    "cluster.routing.allocation.disk.watermark.high": "60%",
    "cluster.routing.allocation.disk.watermark.flood_stage": "70%"
  }
}'
```
4. Get node stats again, they become the same:
```
        "data" : [
          {
            "path" : "/data1/elasticsearch-8.13.0/data",
            "mount" : "/data1 (/dev/vdb)",
            "type" : "ext4",
            "total_in_bytes" : 21003583488,
            "free_in_bytes" : 18922033152,
            "available_in_bytes" : 18261012480,
            "effective_low_watermark" : "50%",
            "effective_high_watermark" : "60%",
            "effective_flood_stage" : "70%"
          }
        ],
...
        "data" : [
          {
            "path" : "/data1/elasticsearch-8.13.0/data",
            "mount" : "/data1 (/dev/vdb)",
            "type" : "ext4",
            "total_in_bytes" : 21003583488,
            "free_in_bytes" : 18706501632,
            "available_in_bytes" : 18045480960,
            "effective_low_watermark" : "50%",
            "effective_high_watermark" : "60%",
            "effective_flood_stage" : "70%"
          }
        ],
```

So this should be expected? Or could you please guide me in the right direction?

DaveCTurner replied:
The issue is that all the allocation decisions are made on the elected master, using the elected master's watermark configuration, so we must ignore the values from `elasticsearch.yml` on all other nodes. Your example shows that we're not ignoring those values here.

howardhuanghua replied:
Got it, so we should always get threshold settings from `clusterSettings` right?

DaveCTurner replied:
No, `clusterSettings` will fall back to the local node's `elasticsearch.yml` if the watermarks are not set in the cluster state which isn't right. We need to reach out to the elected master here. I think I'd extend `org.elasticsearch.action.admin.cluster.allocation.TransportGetAllocationStatsAction` to include this information in its response, since we already invoke this action as part of the node-stats API, but in principle we could add another action just for this info too.

howardhuanghua replied:
Okay, I am going to extend `TransportGetAllocationStatsAction`.

howardhuanghua replied:
Hi @DaveCTurner, I have checked https://github.com/elastic/elasticsearch/pull/105894, so it's better to add threshold watermarks into `NodeAllocationStats`? If user requires `allocations` metric in `_node/stats` then we return threshold watermark stats in this allocations blob?

DaveCTurner replied:
> it's better to add threshold watermarks into `NodeAllocationStats`? 

The exact watermarks depend on both the watermark config (from the elected master) and the actual disk sizes (from the individual nodes). I think it'd be better to include just the watermark config in `org.elasticsearch.action.admin.cluster.allocation.TransportGetAllocationStatsAction.Response` and then have the node coordinating the node stats request combine this information with the disk sizes of the individual nodes according to each node's `FsInfo`.

> If user requires `allocations` metric in `_node/stats` then we return threshold watermark stats in this allocations blob?

I think I preferred your original idea of including this in each node's `FsInfo` section, which is returned if the user requests the `fs` metric.

DaveCTurner commented on server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:
```diff
@@ -36,6 +36,10 @@ public static class Path implements Writeable, ToXContentObject {
         long free = -1;
         long available = -1;
 
+        String effectiveLowWatermark = null;
+        String effectiveHighWatermark = null;
+        String effectiveFloodStage = null;
```
Let's keep these as numbers of bytes, not strings. We can convert them to percentages when rendering a response if needed.

howardhuanghua replied:
If we want to render thresholds as percent, we also need to save a total disk space in bytes?
For example: `number of watermark bytes / total disk space = x%`

DaveCTurner commented on server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:
```diff
@@ -56,6 +60,9 @@ public Path(StreamInput in) throws IOException {
             total = in.readLong();
             free = in.readLong();
             available = in.readLong();
+            effectiveLowWatermark = in.readOptionalString();
+            effectiveHighWatermark = in.readOptionalString();
+            effectiveFloodStage = in.readOptionalString();
```
This is a change to the wire protocol so needs a new `TransportVersion` and some way to handle their omission when talking to older nodes.

DaveCTurner reviewed (changes requested):
Thanks @howardhuanghua, I left some initial comments.

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

howardhuanghua commented:
Hi @DaveCTurner , I have fixed comment discussed issues, please help to review again.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetAllocationStatsAction.java:
```diff
@@ -63,21 +65,22 @@ public TransportGetAllocationStatsAction(
         if (clusterService.state().getMinTransportVersion().before(TransportVersions.ALLOCATION_STATS)) {
             // The action is not available before ALLOCATION_STATS
-            listener.onResponse(new Response(Map.of()));
+            listener.onResponse(new Response(Map.of(), diskThresholdSettings));
```
I think we should be passing back `null` in the bwc case?

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetAllocationStatsAction.java:
```diff
@@ -63,21 +65,22 @@ public TransportGetAllocationStatsAction(
     @Override
     protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<Response> listener) throws Exception {
-        listener.onResponse(new Response(allocationStatsService.stats()));
+        listener.onResponse(new Response(allocationStatsService.stats(), diskThresholdSettings));
```
I would prefer to check `clusterService.state().getMinTransportVersion()` against `TransportVersions.WATERMARK_THRESHOLDS_STATS` here and pass a `null` if the cluster is not fully upgraded. If we don't, then we potentially drop this value on the transport layer (which tends to be trappy), but even if the values get all the way to the user on this request they may not on the next request and that is confusing and not in line with [these docs](https://www.elastic.co/guide/en/elastic-stack/8.13/upgrading-elasticsearch.html#rolling-upgrades):

> During a rolling upgrade, the cluster continues to operate normally. However, any new functionality is disabled or operates in a backward compatible mode until all nodes in the cluster are upgraded. New functionality becomes operational once the upgrade is complete and all nodes are running the new version.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetAllocationStatsAction.java:
```diff
@@ -110,23 +113,35 @@ public ActionRequestValidationException validate() {
     public static class Response extends ActionResponse {
 
         private final Map<String, NodeAllocationStats> nodeAllocationStats;
+        private DiskThresholdSettings diskThresholdSettings = null;
```
I'd rather this was `final` and assigned on every path through the constructors, and also annotated as `@Nullable`:

```suggestion
        @Nullable // for bwc
        private final DiskThresholdSettings diskThresholdSettings;
```

DaveCTurner replied:
This one is still outstanding

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetAllocationStatsAction.java:
```diff
@@ -110,23 +113,35 @@ public ActionRequestValidationException validate() {
             out.writeMap(nodeAllocationStats, StreamOutput::writeString, StreamOutput::writeWriteable);
+            if (out.getTransportVersion().onOrAfter(TransportVersions.WATERMARK_THRESHOLDS_STATS)) {
+                diskThresholdSettings.writeTo(out);
+            }
```
If we check `clusterService.state().getMinTransportVersion()` against `TransportVersions.WATERMARK_THRESHOLDS_STATS` as I suggested above then we can assert that `diskThresholdSettings == null` in the `else` branch here, good for future readers.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetAllocationStatsAction.java:
```diff
@@ -110,23 +113,35 @@ public ActionRequestValidationException validate() {
             return nodeAllocationStats;
         }
+
+        public DiskThresholdSettings getDiskThresholdSettings() {
```
```suggestion
        @Nullable // for bwc
        public DiskThresholdSettings getDiskThresholdSettings() {
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodeStats.java:
```diff
@@ -180,7 +182,7 @@ public NodeStats withNodeAllocationStats(@Nullable NodeAllocationStats nodeAlloc
             jvm,
             threadPool,
-            fs,
+            masterThresholdSettings == null ? fs : fs.renderEffectiveWaterMark(masterThresholdSettings),
```
Could we move this `null` check inside `renderEffectiveWaterMark()`?

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java:
```diff
@@ -199,6 +207,31 @@ public DiskThresholdSettings(Settings settings, ClusterSettings clusterSettings)
         clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_ALLOCATION_DISK_THRESHOLD_ENABLED_SETTING, this::setEnabled);
     }
 
+    public static DiskThresholdSettings readFrom(StreamInput in) throws IOException {
+        DiskThresholdSettings diskThresholdSettings = new DiskThresholdSettings();
+        diskThresholdSettings.setLowWatermark(RelativeByteSizeValue.readFrom(in));
```
I have a slight preference for not adding a new empty constructor here and instead constructing a suitable `Settings` (and empty `ClusterSettings`) to pass into the existing constructor.

DaveCTurner replied:
I would also be ok with adding a new constructor that accepts all the values at once, I just don't think we should allow constructing a `DiskThresholdSettings` with null values for its fields.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/unit/RelativeByteSizeValue.java:
```diff
@@ -103,4 +108,27 @@ public String getStringRep() {
+            absolute.writeTo(out);
+        } else {
+            assert ratio != null;
+            out.writeDouble(ratio.getAsPercent());
```
Could we make `RatioValue implements Writeable` too rather than relying on knowing that `getAsPercent` is a faithful representation of the value on the wire?

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/unit/RelativeByteSizeValue.java:
```diff
@@ -103,4 +108,27 @@ public String getStringRep() {
+    }
+
+    public static RelativeByteSizeValue readFrom(StreamInput in) throws IOException {
+        RelativeByteSizeValue relativeByteSizeValue;
+        boolean isAbsolute = in.readBoolean();
+        if (isAbsolute) {
+            relativeByteSizeValue = new RelativeByteSizeValue(ByteSizeValue.readFrom(in));
+        } else {
+            relativeByteSizeValue = new RelativeByteSizeValue(new RatioValue(in.readDouble()));
+        }
+        return relativeByteSizeValue;
```
nit: I'd suggest inlining these local variables:

```suggestion
        if (in.readBoolean()) {
            return new RelativeByteSizeValue(ByteSizeValue.readFrom(in));
        } else {
            return new RelativeByteSizeValue(RatioValue.readFrom(in));
        }
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/unit/RelativeByteSizeValue.java:
```diff
@@ -9,14 +9,19 @@
  * 3. Relative ratio value (0.95)
  */
-public class RelativeByteSizeValue {
+public class RelativeByteSizeValue implements Writeable {
```
Could we add some tests to `org.elasticsearch.common.unit.RelativeByteSizeValueTests` to show that this round-trips over the wire correctly?

DaveCTurner commented on server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:
```diff
@@ -36,6 +38,11 @@ public static class Path implements Writeable, ToXContentObject {
         long free = -1;
         long available = -1;
 
+        ByteSizeValue effectiveLowStage = null;
+        ByteSizeValue effectiveHighStage = null;
+        ByteSizeValue effectiveFloodStage = null;
+        ByteSizeValue effectiveFrozenFloodStage = null;
```
naming nit/suggestion (it's hard to remember whether these things are the free bytes or the used bytes otherwise)

```suggestion
        ByteSizeValue lowStageFreeBytes = null;
        ByteSizeValue highStageFreeBytes = null;
        ByteSizeValue floodStageFreeBytes = null;
        ByteSizeValue frozenFloodStageFreeBytes = null;
```

DaveCTurner replied:
I realised when you did this that I didn't like the names I had suggested 🤦 see c5a72ea647a6660b7d8fead74fb199653fbd6e63 for my preference.

DaveCTurner commented on server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:
```diff
@@ -92,6 +99,50 @@ public ByteSizeValue getAvailable() {
             return ByteSizeValue.ofBytes(available);
         }
 
+        public void calcMasterBaseEffectiveStage(final DiskThresholdSettings masterThresholdSettings) {
```
naming nit/suggestion:

```suggestion
        public void setEffectiveWatermarks(final DiskThresholdSettings masterThresholdSettings) {
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:
```diff
@@ -146,6 +201,19 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                 builder.humanReadableField(Fields.AVAILABLE_IN_BYTES, Fields.AVAILABLE, getAvailable());
             }
 
+            if (effectiveLowStage != null) {
+                builder.field(Fields.EFFECTIVE_LOW_WATERMARK, getEffectiveLowWatermarkPercent());
+            }
+            if (effectiveHighStage != null) {
+                builder.field(Fields.EFFECTIVE_HIGH_WATERMARK, getEffectiveHighWatermarkPercent());
+            }
+            if (effectiveFloodStage != null) {
+                builder.field(Fields.EFFECTIVE_FLOOD_STAGE, getEffectiveFloodWatermarkPercent());
+            }
+            if (effectiveFrozenFloodStage != null) {
+                builder.field(Fields.EFFECTIVE_FROZEN_FLOOD_STAGE, getEffectiveFrozenFloodWatermarkPercent());
+            }
```
I'd rather we emitted the bytes values instead of the percentages here (using `builder.humanReadableField`)

DaveCTurner replied:
... and the labels should be something like `low_watermark_free_space` (for the human-readable field) and `low_watermark_free_space_in_bytes` (for the raw number) etc.

DaveCTurner commented on server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:
```diff
@@ -460,6 +528,13 @@ public FsInfo(StreamInput in) throws IOException {
         this.total = total();
     }
 
+    public FsInfo renderEffectiveWaterMark(final DiskThresholdSettings masterThresholdSettings) {
```
Naming nit/suggestion

```suggestion
    public FsInfo setEffectiveWatermarks(@Nullable final DiskThresholdSettings masterThresholdSettings) {
```

DaveCTurner reviewed (commented):
Looks great overall. I have left some comments but all of them are superficial things. Also you need to run `./gradlew spotlessapply` to clean up some formatting before this will pass CI.

howardhuanghua commented:
Hi @DaveCTurner , I have fixed these superficial things, and also I have added extra checking for dedicated frozen node.
`spotlessapply` has been applied. Thanks for the review.

DaveCTurner reviewed (commented):
One comment remains.

Also unfortunately the CLA checker isn't happy about your first two commits:

```
Author: “danielhuang <“danielhuang>
```

I think the simplest thing here would be to squash them all into one commit (rebased on top of `main`) with the correct author, then force-push to this branch, something like this:

```
git reset --hard main && git merge --squash 253f6e7ea9f6e5bf7b2e75efdfebd3bcaf26e977 && git commit --no-edit && git push --force
```

Normally we don't like force-pushes but we'll cope with it in this one case.

Could you also add a changelog, something like this:

```diff
diff --git a/docs/changelog/107244.yaml b/docs/changelog/107244.yaml
new file mode 100644
index 00000000000..f805796674f
--- /dev/null
+++ b/docs/changelog/107244.yaml
@@ -0,0 +1,5 @@
+pr: 107244
+summary: Support effective watermark thresholds in node stats API
+area: Allocation
+type: enhancement
+issues: [106676]
```

howardhuanghua committed:
Squashed commit of the following:

commit 253f6e7ea9f6e5bf7b2e75efdfebd3bcaf26e977
Author: howardhuanghua <howard-huang@msn.com>
Date:   Wed Apr 17 22:10:51 2024 +0800

    fix comments

commit f4dd5de494696ffa89eb93a4f23e6ef6252fefd9
Merge: adaea80f668 f5c7938ab80
Author: howardhuanghua <howard-huang@msn.com>
Date:   Wed Apr 17 13:32:02 2024 +0800

    Merge branch 'main' into effective_disk_threshold

commit adaea80f66810866a7ad55ac65c576026a9d09b0
Author: howardhuanghua <howard-huang@msn.com>
Date:   Wed Apr 17 13:13:23 2024 +0800

    update effective thresholds based on elected master node settings.

commit e4e08c897b0d57d4c4e495f57a1f4d9b70e3757f
Merge: 8c047ebc410 d3e0ec157dc
Author: “danielhuang <“danielhuang>
Date:   Sun Apr 14 17:39:25 2024 +0800

    Merge branch 'main' into effective_disk_threshold

commit 8c047ebc4101b41cd4d5c2e6037c313089078203
Author: “danielhuang <“danielhuang>
Date:   Tue Apr 9 09:20:27 2024 +0800

    Support effective watermark thresholds in node stats API

howardhuanghua committed:
Add changelog

howardhuanghua commented:
Thank @DaveCTurner , everything should be fine now.

DaveCTurner committed:
Remove obsolete comments

DaveCTurner committed:
Fix up duff naming suggestion

DaveCTurner committed:
Push branch down

DaveCTurner committed:
final

DaveCTurner committed:
Less branchy

DaveCTurner committed:
Inline & use var

DaveCTurner committed:
Try-with-resources

DaveCTurner commented:
@elasticmachine test this please

DaveCTurner commented:
Looks good - I cleaned up a few things not worth bothering with a separate review for, please check the commits I just pushed and revert or amend any you don't like.

DaveCTurner committed:
Handle null `fsInfo`

DaveCTurner commented:
@elasticmachine test this please

DaveCTurner reviewed (approved):
CI seems happy so LGTM as of 01b2f84ffe9e602c97d433757ea2d3cdc6893770 so I plan on merging this tomorrow morning (UK time).

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/TransportNodesStatsAction.java:
```diff
@@ -90,17 +91,24 @@ protected void newResponseAsync(
                 TransportGetAllocationStatsAction.TYPE,
```
Ah I just realised we need to run this action if the user requests either `allocations` or `fs` now, otherwise we won't get the full `FsInfo`. I think we're missing a test here, something like `rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/nodes.stats/80_allocation_stats.yml` but for the `fs` metric.

howardhuanghua replied:
Thanks for your efforts. Added rest `fs` metric test case.

DaveCTurner reviewed (changes requested):
Sorry just spotted a small problem here, could you take a look?

howardhuanghua committed:
Add fs rest test case

DaveCTurner committed:
Introduce NodeFeature

DaveCTurner reviewed (approved):
We've changed how we do version skips in YAML tests recently so I've adapted things in 8dfec4911bc30bc57d48def858a7b24743635422 (and also fixed the test by adding the check for the `FS` metric). LGTM now.

DaveCTurner commented:
@elasticmachine test this please

DaveCTurner committed:
Avoid forbidden API

DaveCTurner committed:
Nullable on the wire

DaveCTurner commented:
@elasticmachine test this please

DaveCTurner committed:
Update META-INF
